src/bentoml/_internal/client/__init__.py

```diff
@@ -0,0 +1,161 @@
+        try:
+            from .http import HTTPClient
+
+            HTTPClient.wait_until_server_ready(host, port, timeout, **kwargs)
+        except BadStatusLine:
+            # when address is a RPC
+            from .grpc import GrpcClient
+
+            GrpcClient.wait_until_server_ready(host, port, timeout, **kwargs)
+        except Exception as err:
+            # caught all other exceptions
+            logger.error("Failed to connect to server %s:%s", host, port)
+            logger.error(err)
+            raise
+
+    @t.overload
+    @staticmethod
+    def from_url(
+        server_url: str, *, kind: None | t.Literal["auto"] = ...
+    ) -> GrpcClient | HTTPClient:
+        ...
+
+    @t.overload
+    @staticmethod
+    def from_url(server_url: str, *, kind: t.Literal["http"] = ...) -> HTTPClient:
+        ...
+
+    @t.overload
+    @staticmethod
+    def from_url(server_url: str, *, kind: t.Literal["grpc"] = ...) -> GrpcClient:
+        ...
+
+    @staticmethod
+    def from_url(
+        server_url: str, *, kind: str | None = None, **kwargs: t.Any
+    ) -> Client:
+        if kind is None or kind == "auto":
+            try:
+                from .http import HTTPClient
+
+                return HTTPClient.from_url(server_url, **kwargs)
+            except BadStatusLine:
+                from .grpc import GrpcClient
+
+                return GrpcClient.from_url(server_url, **kwargs)
+            except Exception as e:  # pylint: disable=broad-except
+                raise BentoMLException(
+                    f"Failed to create a BentoML client from given URL '{server_url}': {e} ({e.__class__.__name__})"
+                ) from e
+        elif kind == "http":
+            from .http import HTTPClient
+
+            return HTTPClient.from_url(server_url, **kwargs)
+        elif kind == "grpc":
+            from .grpc import GrpcClient
+
+            return GrpcClient.from_url(server_url, **kwargs)
+        else:
+            raise BentoMLException(
+                "Invalid client kind. Must be one of ['http', 'grpc', 'auto']"
+            )
+
+    def _sync_call(
+        self, inp: t.Any = None, *, _bentoml_api: InferenceAPI, **kwargs: t.Any
+    ):
+        return asyncio.run(self._call(inp, _bentoml_api=_bentoml_api, **kwargs))
+
+    @abstractmethod
+    async def _call(
+        self, inp: t.Any = None, *, _bentoml_api: InferenceAPI, **kwargs: t.Any
+    ) -> t.Any:
+        raise NotImplementedError
+
+    def __enter__(self):
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> bool | None:
+        pass
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> bool | None:
+        pass
```

Comment:
What does the context manager do for us again?

Reply:
Currently nothing, but ideally it should be used for creating channel and closing it after.

Reply:
I guess we can leave this in, not much point removing it.